MSC2918: Refresh tokens (#2918)
matrix-org/matrix-doc: proposals/2918-refreshtokens.md

# MSC2918: Refresh tokens

In Matrix, requests to the Client-Server API are currently authenticated using non-expiring, revocable access tokens.
An access token might leak for various reasons, including:

 - leaking from the server database (and its backups)
 - intercepting it with a man-in-the-middle attack
 - leaking from the client storage (and its backups)

In the OAuth 2.0 world, this vector of attack is partly mitigated by having expiring access tokens with short lifetimes and rotating refresh tokens to renew them.

> This MSC adds support for expiring access tokens and introduces refresh tokens to renew them.

Reviewer: ... why?

I'm a bit confused as to how this made it to FCP, without the proposal even specifying *how* this new mechanism is supposed to mitigate credential leaks.

One way or another, the device will need to hold on to *some* sort of long-lived credential, whether you call it an 'access token' or a 'refresh token' or anything else, so how is this not just more steps to achieve the same level of security? Under what specific threat model does this mitigate an attack, and how?

Reviewer 2: Neither the access token, nor the refresh token is long lived. The refresh token is single use and the access token is time limited. As such they both are only useable for a limited time, if they get leaked (i.e. someone posts their curl command in #synapse:matrix.org or so). I would still like to see some more numbers on if this actually helps or is just done "because oauth does it".

Also, this is not in FCP, it just has FCP proposed.

Reviewer: > Neither the access token, nor the refresh token is long lived.

With "long-lived credential", I am not referring to any one specific token, but rather the more abstract state of the client having *some* kind of way to authenticate a particular session, even if a specific token is rotated. This is typically what refresh tokens are claimed to prevent, but they don't really, at least not in the general case - one way or another, the client has long-lived access to a session, and so compromising the client means compromising that session.

> The refresh token is single use and the access token is time limited. As such they both are only useable for a limited time, if they get leaked (i.e. someone posts their curl command in #synapse:matrix.org or so). I would still like to see some more numbers on if this actually helps or is just done "because oauth does it".

I agree with this, and IMO this is why it's a problem that the MSC doesn't set out the specific mitigation impact. It's something that *might* have value in a limited set of circumstances, but it does actually need to be motivated and shown to be worth the complexity - complexity in security-critical code increases attack surface, and so ineffective security measures can actually end up making a system *less* secure, even if you ignore the practical hassle of implementing that complexity.

In other words, the MSC should be very explicit about:
1. What specific attacks this is meant to protect from under which specific circumstances
2. How meaningful that is in the bigger picture
3. How that justifies the additional complexity

> Also, this is not in FCP, it just has FCP proposed.

Ah, sorry, my bad. My concern remains the same, though - this is a security-critical proposal and it's concerning to me that people perceive it as "looks good" when it's still missing crucial rationale.

Author: @joepie91: Thanks a lot for your remarks! I agree that this MSC might need another round of clarification around why this change makes sense. That being out of the way, let me list a few scenarios where this change is useful and IMO justifies the added complexity.

It does not quite protect against compromised client, but while I'm not saying it is expected, keep in mind that right now compromised clients can do a lot of bad things since you usually enter your account credentials directly in the client and nothing prevents them to record those credentials, send them somewhere else, and complete UIA-protected operations on your behalf without you noticing.

In contrast, it protects against a whole range of attacks where the access and/or refresh token get leaked but the client isn't completely compromised.

For example, those token can leak from user backups (user backs up his device on a NAS, the NAS gets compromised and leaks a backup of the clients secret storage), but we can assume those backups could be at least 5 min old.
If the leak only includes the access token, it is useless to the attacker since it would have expired ; if it also includes the refresh token, it is useless *if* the token was refreshed before (which will happen if the user just opens their Matrix client in between).
Worst case scenario, the leaked refresh token is still valid: in this case, the attacker would consume the refresh token to get a valid access token, but when the original client tries to use the same refresh token, the homeserver can detect it, consider the session has been compromised, end the session and warn the user.

This kind of attack also applies to leakage from the server, which could happen from database backups for example.

The important thing here is while it does not completely prevent attacks in case of a token leakage, it does make this range of attack a lot more time-sensitive and detectable (a HS will notice if a refresh token is being used twice).

Guidelines around [refresh tokens from the IETF](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#section-4.13.2) are IMO pretty interesting. Side note here, this document is pretty interesting overhaul and quite worth reading if you have an OAuth-like auth stack. They recommend that either:

 - the refresh token are sender-bound and require client authentication (making token leakage completely useless if the client credentials are not leaked at the same time)
 - or make them rotate to make the attack a lot harder as described just above.

Since all clients are basically "public" in the Matrix world, we don't have client-bound credentials we could use, hence the rotation of refresh tokens.

---

The other kind of scenario where this change makes sense is to help further changes in the homeservers.
A good, recent example of this, is in Synapse v1.34.0 [they moved away from macaroons for access tokens](https://github.com/matrix-org/synapse/pull/5588) to random, shorter, saved in database tokens, similar to [what GitHub did recently](https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/).
Because there is no refresh token mechanism in the C2S API, most Synapse instances now have a mix of the two formats of tokens, and for a long time. It makes it impossible to enforce the new format of tokens without invalidating all existing sessions, making it impossible to rollout stuff like a web-app firewall in front of Synapse that verifies the shape and checksums of tokens even before reaching Synapse.

---

Last but not least, expiring tokens is already a thing that exists in Synapse (via the `session_lifetime` parameter). Before this MSC, clients had no idea when the session would end and relied on the server replying with a 401 error with `soft_logout: true` in the response on a random request to trigger a soft logout and go through the auth process again.
The nice side-effect of this MSC (although it could have been introduced separately) is that the login responses can now include a `expires_in_ms` to inform the clients when the token will expire.

---

> In other words, the MSC should be very explicit about: …

I did not include those detailed scenarios in the MSC itself to keep it short, but in retrospect I should have had this at least in the original PR comment. As said earlier, I'm happy to include changes that clarify the MSC purpose if you think the current version lacks some explanation.
I'm quite curious of what other members of the spec core team expect in terms of details and rationale in a MSC.

I will also follow up in the coming week on how this fits in the bigger picture, although I'm convinced that this change is a step in the right direction in terms of security and does improve the auth stack by itself regardless of future evolutions.

Reviewer 2: @sandhose Thanks a lot for the extensive explanation. I am very interested in the bigger picture too! If we are already changing stuff in breaking ways, it probably makes sense to have the bigger picture so that we can bundle up the breaking changes in the spec (even if an MSC is passed at an earlier date, see spaces for example, which got approved but will probably not be specced in the next spec release, so that all the changes for it can land in one release together). Since this makes a lot of stuff harder too, as security stuff usually does, it would be nice to see the solutions for bots and such as well. I'm also still partial to UIA for /login, although I know that is blocked on the oauth experiments. So again, thank you a lot for the write up and looking forward to know more about your plans, even if that is not required for this MSC probably!

Reviewer 3: >  I'm quite curious of what other members of the spec core team expect in terms of details and rationale in a MSC.

I can't necessarily speak on behalf of the *whole* SCT, but:

Some explanation of the rationale behind an MSC is absolutely expected. https://spec.matrix.org/unstable/proposals/#process says: "clearly state the problem being solved". That is open to interpretation, and I think that's no bad thing. The amount of explanation and justification should vary according to the MSC. If the MSC is an unambiguous improvement, then we need say very little. If the change is more controversial, that suggests we need to do better at justifying it.

In this case, I think much of your excellent comment above could be incorporated in the MSC document wholesale.

I'd recommend adding a separate "Rationale" section at the bottom so that people who want to get straight to the technical details can do so.

Reviewer 4: Added justification for why we're not doubling down on UIA would be appreciated, particularly for the community perspective. Internally (and I believe a multitude of times publicly) we've talked about how UIA is deficient in delivering the functionality we want, such as by reinventing the wheel or being fairly difficult to work within for the more complex stages. 

It may be worth mentioning that down the line we may end up with an MSC that changes all of UIA over to OAuth 2.0, but for now this is effectively changing the token format and not the entry points into the authentication system: submitting usernames and passwords (or using SSO providers) is still a very capable thing.

----

@deepbluev7 fwiw, I don't see the bots usecase as a very strong one to keeping the access tokens as-is. Bots can update, as can scripts. A number of libraries already supply the functionality and scripts would easily be able to call the renew endpoint. 

For scripts in particular: an approach would be to persist the *refresh* token where it was previously persisting the access token. This does mean that the script might need a form of persistence (not impossible for a script), but still leaves the scripts highly atomic. Because they'd be persisting the refresh token, the first thing they'd do is exchange it for an access token and store the new refresh token. This also means that if something went wrong with the job then it'd be safe to re-run (auth-wise, at least) because the refresh token would simply award a new access token. Alternatively, the script could take a username and password, and hit `/logout/all` when it's done - this is effectively as good as the old-style access token, though does have the higher risk of being able to complete UIA stages (see also: replacing with OAuth 2.0) autonomously. 

Other authentication methods, including custom ones which emulate the old-style access tokens, may also be appropriate. Similarly, servers which are told that the access/refresh token will be for a script/bot could simply expose a 999 year expiration on that token. Taking these measures have security impact to consider though, as it's obvious Matrix is switching to a shorter-lived approach for a reason.

Reviewer 2: @turt2live Persisting an access token is not really practical in a CI environment in my opinion. CI pipelines should not be stateful, as one mistake means the pipeline is now broken for everyone until an admin manually fixes it. I use scripts in my CI pipeline to upload nightlies and send notifications. So that needs something else than refresh tokens. I agree that increasing the rotation period for that specific token or fetching some other token would be possible, but storing the password is a horrible solution and such special bot tokens should not need more privileges than the current auth tokens, otherwise some bots can only be run by an admin now. I don't want to keep the current mechanism as it is just for bots, but I want a solution for bots (that can't persist state) before all matrix servers require refresh tokens.

Reviewer 5: Fwiw I think linking to some standard copy of refresh tokens are a god idea (eg. the oauth draft section you linked) and a couple of sentences answer the easiest questions and spelling out the things you go on to say later would go a long way, eg. "Refresh tokens can be used only once, so ensure a leaked access token cannot continue to be used. Idempotency is achieved by only invalidating the refresh token after the new access token is used. See <link> for more on refresh tokens as a general concept".

The detailed explanation here is fantastic but whatever doesn't make it into the MSC itself will more-or-less be lost to history.

- `refresh_token`: The refresh token, which can be used to obtain new access tokens.

This also applies to logins done by application services.

Both fields are optional.
If `expires_in_ms` is missing, the client can assume the access token won't expire.
If `refresh_token` is missing but `expires_in_ms` is present, the client can assume the access token will expire but it won't have a way to refresh the access token without re-logging in.

### Account registration API changes

Unless `inhibit_login` is `true`, the account registration API returns two additional fields:

- `expires_in_ms`: The lifetime in milliseconds of the access token.
- `refresh_token`: The refresh token, which can be used to obtain new access tokens.

This also applies to registrations done by application services.

As in the login API, both field are optional.

### Token refresh API

This API lets the client refresh the access token.
A new refresh token is also issued.
The existing refresh token remains valid until the new access token (or refresh token) is used, at which point it is revoked.
This allows for the request to get lost in flight.

> The Matrix server can revoke the old access token right away, but does not have to since its lifetime is short enough that it will expire anyway soon after.

Reviewer 6: > ... but does not have to since its lifetime is short enough that it will expire anyway soon after.

Is this necessarily true?  Is there a requirement that clients only refresh the access token shortly before the old one expires?

I would think that it would be best to say that the server must revoke the old access token once the new access/refresh token is used.

Reviewer 7: It seems as though the MSC now says this -- If so, @sandhose note that it's best to resolve outstanding PR comments once they've been addressed.

Reviewer 6: The point is that a server could set a lifetime for the access token to, e.g. 2 weeks, and a client could refresh the token 5 minutes after receiving it.  Then the old access token would be valid for 2 weeks minus 5 minutes, which I don't think counts as it "expir[ing] ... soon after".  So this statement seems to be incorrect.

Reviewer 6: Ah, I see that the MSC now says that access token expires after a short amount of time, so I think this is resolved now.

}
```

If the `refresh_token` is missing from the response, the client can assume the refresh token has not changed and use the same token in subsequent token refresh API requests.

The `refresh_token` parameter can be invalid for two reasons:

 - if it does not exist
 - if it was already used once

In both cases, the server must reply with a `401` HTTP status code and an `M_UNKNOWN_TOKEN` error code.
This new use case of the `M_UNKNOWN_TOKEN` error code must be reflected in the spec.
As with other endpoints, the server can include an extra `soft_logout` parameter in the response to signify the client it should do a soft logout.

This new API should be rate-limited and does not require authentication since only the `refresh_token` parameter is needed.
Identity assertion via the `user_id` query parameter as defined by the Application Service API specification is disabled on this endpoint.

### Device handling

The current spec states that "Matrix servers should record which device each access token is assigned to".
This must be updated to reflect that devices are bound to a session, which are created during login and stays the same after refreshing the token.

> 

Reviewer 6: Should there be a new error code for indicating that the access token has expired and needs to be refreshed?

I think this might be useful for CLI tools that might want to be fairly dumb. e.g. they can just hit the endpoint with the current access token, and if it says that the token has expired, then refresh the token and try again.  This way, it doesn't have to worry about keeping track of time.

Or it could be helpful for situations where a client's clock gets messed up and it thinks that the access token should still be valid when it actually isn't.

Reviewer 7: Clients probably should be keeping track of time anyways, otherwise you'll find your client wastes a request every 5m to find out that the token has expired. IMO keeping track of time isn't too hard - even embedded systems do it quite regularly.

> Or it could be helpful for situations where a client's clock gets messed up and it thinks that the access token should still be valid when it actually isn't.

This is a more convincing point. If the token is invalid and the clock says that a refresh shouldn't be necessary, the client could assume that it needs to re-login again. Though, perhaps the client should attempt a refresh of the token anyways, in which case this issue would be sidestepped...

Whatever the conclusion, probably warrants a paragraph in the "Potential issues" or "Alternatives" sections.

Reviewer 6: Right.  I think that part of the change that clients will need to make is, instead of assuming that they're logged out if they get an `M_UNKNOWN_TOKEN`, now if they get an `M_UNKNOWN_TOKEN`, they should try to refresh their token, and if that doesn't work, then they're logged out.  I think this needs to be clearer in the MSC (and also how this interacts with Soft logout).

Reviewer 6: Maybe add something like this to the "Proposal" section:

> If a client uses an access token that has expired, the server will respond with an `M_UNKNOWN_TOKEN` error.  Thus, if a client receives an `M_UNKNOWN_TOKEN` error, and it has a refresh token available, it should no longer assume that it has been logged out, and instead attempt to refresh the token using the new `/refresh` endpoint.  If the client was in fact logged out, then the server will respond with an `M_UNKNOWN_TOKEN` error, possibly with the `soft_logout` parameter set.

Author: Thanks, added that in https://github.com/matrix-org/matrix-doc/pull/2918/commits/c0767632c83e97f7af3204525e73953de0268c16

## Potential issues

The refresh token being rotated on each refresh is strongly recommended in the OAuth 2.0 world for unauthenticated clients to avoid token replay attacks.
This can however make the deployment of CLI tools for Matrix a bit harder, since the credentials can't be statically defined anymore.
This is not an issue in OAuth 2.0 because usually CLI tools use the client credentials flow, also known as service accounts.
An alternative would be to make the refresh token non-rotating for now but recommend clients to support rotation of refresh tokens and enforce it later on.

## Alternatives

This MSC defines a new endpoint for token refresh, but it could also be integrated as a new authentication mechanism.

## Security considerations

The time to live (TTL) of access tokens isn't enforced in this MSC but is advised to be kept relatively short.
Servers might choose to have stateless, digitally signed access tokens (JWT are good examples of this), which makes them non-revocable.
The TTL of access tokens should not exceed 15 minutes if they are revocable and 5 minutes if they are not.

## Unstable prefix

> While this MSC is not in a released version of the specification, clients should add a `org.matrix.msc2918.refresh_token=true` query parameter on the login endpoint, e.g. `/_matrix/client/r0/login?org.matrix.msc2918.refresh_token=true`.

Reviewer 6: ```suggestion
While this MSC is not in a released version of the specification, clients should add a `org.matrix.msc2918.refresh_token=true` query parameter on the login and registration endpoints, e.g. `/_matrix/client/r0/login?org.matrix.msc2918.refresh_token=true`.
```

?

> While this MSC is not in a released version of the specification, clients should add a `org.matrix.msc2918.refresh_token=true` query parameter on the login endpoint, e.g. `/_matrix/client/r0/login?org.matrix.msc2918.refresh_token=true`.

Reviewer 7: I think this parameter signals support from the client for handling refresh tokens, however there is no mention anywhere of what one should do when unstable prefixes can be removed. Will a parameter `/login?refresh_token=true` stick around after we remove unstable prefixes, or will the parameter be removed altogether as it is assumed clients will be upgraded by then?

Author: This has been replaced by a field in the request body, that will stick once the MSC is released, see https://github.com/matrix-org/matrix-doc/pull/2918/commits/4cf821c7c2339b6a46676c0a8b92b1593e8850a9

# MSC2918: Refresh tokens

In Matrix, requests to the Client-Server API are currently authenticated using non-expiring, revocable access tokens.
An access token might leak for various reasons, including:

 - leaking from the server database (and its backups)
 - intercepting it with a man-in-the-middle attack
 - leaking from the client storage (and its backups)

In the OAuth 2.0 world, this vector of attack is partly mitigated by having expiring access tokens with short lifetimes and rotating refresh tokens to renew them.
This MSC adds support for expiring access tokens and introduces refresh tokens to renew them.
A more [detailed rationale](#detailed-rationale) of what kind of attacks it mitigates lives at the end of this document.

## Proposal

> The access token returned by the login endpoint expires after a short amount of time, forcing the client to renew it with a refresh token.

Reviewer 6: Can we have an example of what is considered a "short" amount of time?

Reviewer 8: It's stated in a couple of places (one just a few lines below, and another in Security considerations) - from those, "short" means "several minutes, could be more if the token is revocable"?

Reviewer 6: I saw the mention a couple lines below, but that seems to be distinct from this as it's specifically about non-revocable tokens.  But I now see the mention in Security Considerations, which answers my question.

The existing refresh token remains valid until the new access token (or refresh token) is used, at which point it is revoked.
This allows for the request to get lost in flight.
The Matrix server can revoke the old access token right away, but does not have to since its lifetime is short enough that it will expire anyway soon after.

`POST /_matrix/client/r0/refresh`

```json
{
  "refresh_token": "aaaabbbbccccdddd"
}
```

response:

```json
{
  "access_token": "xxxxyyyyzzz",
  "expires_in_ms": 60000,
  "refresh_token": "eeeeffffgggghhhh"
}
```

If the `refresh_token` is missing from the response, the client can assume the refresh token has not changed and use the same token in subsequent token refresh API requests.

The `refresh_token` parameter can be invalid for two reasons:

 - if it does not exist
 - if it was already used once

In both cases, the server must reply with a `401` HTTP status code and an `M_UNKNOWN_TOKEN` error code.
This new use case of the `M_UNKNOWN_TOKEN` error code must be reflected in the spec.
As with other endpoints, the server can include an extra `soft_logout` parameter in the response to signify the client it should do a soft logout.

> This new API should be rate-limited and does not require authentication since only the `refresh_token` parameter is needed.

Reviewer 6: I don't know if this is the author's reason, but I would find it confusing if you used different types of tokens in the `Authorization` header for different endpoints.

# MSC2918: Refresh tokens

In Matrix, requests to the Client-Server API are currently authenticated using non-expiring, revocable access tokens.
An access token might leak for various reasons, including:

 - leaking from the server database (and its backups)
 - intercepting it with a man-in-the-middle attack
 - leaking from the client storage (and its backups)

In the OAuth 2.0 world, this vector of attack is partly mitigated by having expiring access tokens with short lifetimes and rotating refresh tokens to renew them.
This MSC adds support for expiring access tokens and introduces refresh tokens to renew them.
A more [detailed rationale](#detailed-rationale) of what kind of attacks it mitigates lives at the end of this document.

## Proposal

> Homeservers can choose to have access tokens returned by the registration and login endpoints after a short amount of time, forcing the client to renew it with a refresh token.

Reviewer 3: ```suggestion
Homeservers can choose to have access tokens expire after a short amount of time, forcing the client to renew them with a refresh token.
```

A refresh token is issued on login and rotates on each usage.

It allows homeservers to opt for signed and non-revocable access tokens (JWTs, Macaroon, etc.) for performance reasons if their expiration is short enough (less than 5 minutes).

> Whether the access token expire and what lifetime they have is up to the homeserver, but client have to support refreshing tokens.

Reviewer 3: @sandhose: is this correct? I think we discussed keeping the query parameter as a way of allowing clients to opt in, even once the MSC lands - mostly as a way of supporting the script usecase until proper OAuth lands.

So something more like:

```suggestion
Clients are not *required* to support refreshing access tokens, though it is heavily recommended.
Clients indicate their support for refreshing access tokens via a query parameter to the `/login` and `/register` APIs.
Homeserver implementations should expire access tokens whenever a client indicates support. 
The lifetime of issued access tokens is up to homeserver deployments, but a lifetime of around 30 minutes (?) would be normal.

Handling of clients that do *not* support refreshing access tokens is up to individual homeserver deployments.
For example, server administrators may choose to support such clients for backwards-compatibility, or to expire access tokens anyway for improved security at the cost of inferior user experience in legacy clients.
```

Author: I opted for an additional field in the request body instead. TTL of access tokens is already mentioned in the "security consideration" part. See https://github.com/matrix-org/matrix-doc/pull/2918/commits/4cf821c7c2339b6a46676c0a8b92b1593e8850a9

Reviewer 6: Thanks.  I think that adding the request body field solves the issue of backward compatibility.  I think this means that this technically doesn't match the implementation, but I think that having the body param is close enough to the existing query param that we don't need to raise the "needs-implementation" flag on this.

Reviewer 5: It still says they're required to support them here though, conflicting with below.

Reviewer 3: yup, what @dbkr said

> ### Login API changes

Reviewer 3: we should probably make it explicit what happens when the access token expires.

On that front: should we recommend that servers set `soft_logout: true` (see https://spec.matrix.org/unstable/client-server-api/#soft-logout) as a way of increasing compatibility with clients that do not support refresh tokens?

Reviewer 3: oh, this is basically the same point as another comment thread (https://github.com/matrix-org/matrix-doc/pull/2918/files#r677648682)

Author: Just did that in https://github.com/matrix-org/matrix-doc/pull/2918/commits/c0767632c83e97f7af3204525e73953de0268c16

If a client uses an access token that has expired, the server will respond with an `M_UNKNOWN_TOKEN` error.
Thus, if a client receives an `M_UNKNOWN_TOKEN` error, and it has a refresh token available, it should no longer assume that it has been logged out, and instead attempt to refresh the token.
If the client was in fact logged out, then the server will respond with an `M_UNKNOWN_TOKEN` error, possibly with the `soft_logout` parameter set.

### Login API changes

The login API returns two additional fields:

- `expires_in_ms`: The lifetime in milliseconds of the access token.
- `refresh_token`: The refresh token, which can be used to obtain new access tokens.

This also applies to logins done by application services.

Both fields are optional.
If `expires_in_ms` is missing, the client can assume the access token won't expire.
If `refresh_token` is missing but `expires_in_ms` is present, the client can assume the access token will expire but it won't have a way to refresh the access token without re-logging in.

Clients advertise their support for refreshing tokens by setting the `refresh_token` field to `true` in the request body.

### Account registration API changes

> Unless `inhibit_login` is `true`, the account registration API returns two additional fields:

Reviewer 9: Shouldn't this be if its *false*? I.e. we include the params when we return a valid access token?

Author: The spec says:

> `inhibit_login`: If true, an access_token and device_id should not be returned from this call, therefore preventing an automatic login. Defaults to false.

So that sentence seems correct? If `inhibit_login` is `true`, it will **not** return the additional fields

Reviewer 3: it's not the easiest thing to grok, but @sandhose is right, and @erikjohnston is confused.

# MSC2918: Refresh tokens

Requests are currently authenticated using non-expiring, revocable access tokens.
This goes against security best practices known in the OAuth2 world.
This MSC make the access tokens expiring and introduces refresh tokens to renew them to fight against token replay attacks.

## Proposal

The access token returned by the login endpoint expires after a short amount of time, forcing the client to renew it with a refresh token.
A refresh token is issued on login and rotates on each usage.

Homeservers can choose to make the access tokens signed and non-revocable for performance reasons if the expiration is short enough (less than 5 minutes).

### Login API changes

> The login API returns two additional fields:

Reviewer 10: How would this be done in a backwards compatible manner? (How would the server know that the client now supports refresh tokens?)

Author: Not sure about that. Maybe a new parameter in the login API?

Author: This was fixed by https://github.com/matrix-org/matrix-doc/pull/2918/commits/f8dad2acd6b554c331793cc642c9155542f5b8c7

- `expires_in`: The lifetime in seconds of the access token.
- `refresh_token`: The refresh token, which can be used to obtain new access tokens.

### Token refresh API

This API lets the client refresh the access token.
A new refresh token is also issued, and the existing one is revoked.
The Matrix server doesn't have to make the old access token invalid, since its lifetime is short enough.

> `POST /refresh`

Reviewer 10: Guessing this is `/_matrix/client/r0/refresh`?

Author: Shouldn't MSCs use an unstable prefix? Not entirely sure how that works

Reviewer 10: They should describe the final API endpoint assuming it is accepted and include unstable information in the "Unstable prefix".

Author: This was fixed by https://github.com/matrix-org/matrix-doc/pull/2918/commits/f8dad2acd6b554c331793cc642c9155542f5b8c7

```json
{
  "refresh_token": "aaaabbbbccccdddd"
}
```

response:

```json
{
  "access_token": "xxxxyyyyzzz",
  "expires_in": 60,
  "refresh_token": "eeeeffffgggghhhh"
}
```

### Device handling

The current spec states that "Matrix servers should record which device each access token is assigned to".

> This must be updated to reflect that devices are bound to a session, which are created during login and stays the same one after refreshing the token.

Reviewer 10: Drawing a conclusion from above:

> The Matrix server doesn't have to make the old access token invalid, since its lifetime is short enough. 

Does this also meaning that a session/device could have multiple access tokens at once?

Author: If the Matrix server chooses to do stateless validation of tokens, yes that would be the case.
For example, if the server uses JWTs as access token, you might have a token that looks like this decoded:

```json
{
  "iat": 1516239022,
  "nbf": 1516239022,
  "exp": 1516239082,
  "iss": "https://matrix.connecteu.rs",
  "mxid": "@quentin:connecteu.rs",
  "device_id": "XXYYZZ"
}
```

But in any cases, clients should not rely on the old access token being still valid I think

# MSC2918: Refresh tokens

Requests are currently authenticated using non-expiring, revocable access tokens.
This goes against security best practices known in the OAuth 2.0 world.
This MSC make the access tokens expiring and introduces refresh tokens to renew them to fight against token replay attacks.

## Proposal

The access token returned by the login endpoint expires after a short amount of time, forcing the client to renew it with a refresh token.
A refresh token is issued on login and rotates on each usage.

Homeservers can choose to make the access tokens signed and non-revocable for performance reasons if the expiration is short enough (less than 5 minutes).

### Login API changes

The login API returns two additional fields:

> - `expires_in`: The lifetime in seconds of the access token.

Reviewer 3: most (all?) time-related fields in Matrix are measured in milliseconds, not seconds. I'd recommend following suit, unless there are compelling reasons to go with seconds, like OAuth conventions.

Author: This is directly inspired from [OAuth 2.0](https://tools.ietf.org/html/rfc6749#section-4.3.3). There isn't a real reason to keep it close to OAuth 2.0 here though, so `expires_in_ms` might make more sense here to keep it consistent

Author: Resolved with https://github.com/matrix-org/matrix-doc/pull/2918/commits/0e615f78757ac2cb8bafacd2eba69deac00f6967

# MSC2918: Refresh tokens

Requests are currently authenticated using non-expiring, revocable access tokens.
This goes against security best practices known in the OAuth 2.0 world.
This MSC make the access tokens expiring and introduces refresh tokens to renew them to fight against token replay attacks.

## Proposal

The access token returned by the login endpoint expires after a short amount of time, forcing the client to renew it with a refresh token.
A refresh token is issued on login and rotates on each usage.

Homeservers can choose to make the access tokens signed and non-revocable for performance reasons if the expiration is short enough (less than 5 minutes).

### Login API changes

The login API returns two additional fields:

- `expires_in_ms`: The lifetime in milliseconds of the access token.
- `refresh_token`: The refresh token, which can be used to obtain new access tokens.

### Account registration API changes

Unless `inhibit_login` is `true`, the account registration API returns two additional fields:

- `expires_in_ms`: The lifetime in milliseconds of the access token.

> - `refresh_token`: The refresh token, which can be used to obtain new access tokens.

Reviewer 3: we should emphasize that this also applies to ASes (see https://matrix.org/docs/spec/application_service/r0.1.2#server-admin-style-permissions): ASes which use the access token returned by that endpoint will need to refresh it.

This also applies to logins done by application services.

### Account registration API changes

Unless `inhibit_login` is `true`, the account registration API returns two additional fields:

- `expires_in_ms`: The lifetime in milliseconds of the access token.
- `refresh_token`: The refresh token, which can be used to obtain new access tokens.

This also applies to registrations done by application services.

### Token refresh API

This API lets the client refresh the access token.
A new refresh token is also issued. The existing refresh token remains valid until the new access token (or refresh token) is used, at which point it is revoked. This allows for the request getting lost in flight.
The Matrix server can but does not have to make the old access token invalid, since its lifetime is short enough.

`POST /_matrix/client/r0/refresh`

```json
{
  "refresh_token": "aaaabbbbccccdddd"
}
```

response:

```json
{
  "access_token": "xxxxyyyyzzz",
  "expires_in_ms": 60000,
  "refresh_token": "eeeeffffgggghhhh"
}
```

The `refresh_token` parameter can be invalid for two reasons:

 - if it does not exist

>  - if it was already used once

Reviewer 9: How does this deal with idempotency? I think it'd be quite annoying, especially on mobile devices, if you had to relogin if the client didn't see the response to the `/refresh_token`. Can we do something like only invalidating the refresh token once we see the new access token be used?

Reviewer 3: that's what the synapse impl does :-p

> # MSC2918: Refresh tokens

Reviewer 3: a few thoughts about backwards-compatibility and breaking changes.

This MSC is a bit unusual in that it proposes something that will definitely break compatibility with old clients, yet isn't technically a breaking change in the API because no existing endpoints/properties/etc are being removed.

I think it's important that the expiry of access tokens is left as optional (to be determined by the server), so that we don't have a hard compatibility break in the spec. So, a client that supported spec `vNext` would *have* to support refresh tokens, though a server that supports spec `vNext` doesn't necessarily have to make use of them.

Then, server implementations/admins can independently decide when they have few enough legacy clients to switch to issuing expiring access tokens by default.

Author: > to switch to issuing expiring access tokens by default.

Isn't it rather "to not support the `v1` API anymore"? Since we can't really enforce expiring token without breaking clients, I expect we'll have to have a new major revision of the API where clients *have to* support refreshing tokens?
Then when enough clients have switched to the `v2` endpoint, admins can choose to not support the `v1` API anymore?

(Sorry if it does not make sense, I'm a bit confused with versions and revisions of the API)

Reviewer 3: > Isn't it rather "to not support the `v1` API anymore"? 

well, if the proposal is to switch to a replacement for `/_matrix/client/r0/login`, the MSC should say that (and the sample implementations should do so rather than using a query-param on `/r0/login`). However, I don't think that is necessary, and indeed makes life harder than it needs to be for client developers.

For clarity: I understand that a successor to `/r0/login` would be called `/_matrix/client/v4/login`, thanks to [MSC2844](https://github.com/matrix-org/matrix-doc/blob/master/proposals/2844-global-versioning.md#endpoint-versioning).

So, suppose we introduce an API `/_matrix/client/v4/login`, and say "if you call this, you MUST support refreshing tokens". If you use that and you are a client, you also have to be prepared to fall back to `/r0/login` (in case you're talking to an older server).  From a client's point of view, you'd rather just hit `/r0/login` and be prepared to deal with new fields in the response.

From a server's point of view, there's nothing stopping the server issuing expiring access tokens today; nor is there anything stopping the server returning the additional fields in the login response - it's just unlikely a client will know what to do with them.

So, the only thing `/v4/login` would bring would be metrics for server admins to know how many clients support refreshing tokens, and it comes at a cost of clients having to implement an annoying fallback (as well as significant complexity in the spec documentation). That seems like a high price to pay, and there are other ways to do it. IMHO `org.matrix.msc2918.refresh_token` is going to be good enough.

Reviewer 6: I agree that this MSC should spell out how backward compatibility is handled, whether by using `/v4/login` (and `/v4/register`), or by adding a parameter (like the `org.matrix.msc2918.refresh_token` parameter mentioned in the unstable prefix).  I think that adding a parameter would be better since it allows clients to just hit `/r0/login`, rather than trying `/v4/login`, seeing if it errors, and then trying `/r0/login` if it does.

Reviewer 6: > > I think that adding a parameter would be better since it allows clients to just hit /r0/login, rather than trying /v4/login, seeing if it errors, and then trying /r0/login if it does.
> 
> As said in another thread, I'd lean towards enforcing refreshing access tokens in the next API version. Wouldn't the clients then be able to know what version the server supports beforehand by checking `/versions`?

I don't think that would be sufficient.  And it's not that we can just say "If `/versions` contains {vNext}, then use refresh tokens".  We'd need to remove all previous versions from the list, as clients that use those versions will break. And a lot of clients (e.g. simple bots) don't check `/versions`, at least for basic functionality, so these would just break.

I think we need an actual plan on how we're going to handle backward compatibility for this.

Reviewer 8: The current approach (with the query parameter for opting in in the meantime, and extra parameters returned in the response) seems to be backwards compatible, no?

Reviewer 6: My understanding is that the the query parameter is only for while this MSC is in development, but at the next spec release, login and registration will emit refresh tokens by default, and old clients that aren't aware of refresh tokens will suddenly break.

FWIW, when I say that we need a "plan", I mean something like "We will shout loudly about this change, and in n months, homeserver implementations can feel free to use refresh tokens by default."  The plan should probably also consider the various Linux distributions.  e.g. Debian bullseye was just released a couple weeks ago, and will probably be the "stable" version for about two years.  We may be able to get updates through backports for clients, but not everyone uses backports.

Reviewer 6: Some possibilities for dealing with backwards compatibility:

- Yell loudly about refresh tokens becoming a thing and warn that they will become default at some point.  Allow clients to specify that they want to use refresh tokens (e.g. by including a new parameter).  After some predefined time, servers can default to using refresh tokens.  Clients can stop setting the new parameter.
  - downsides:
    - clients that haven't been updated will break silently when servers default to using refresh tokens.
    - if clients stop setting the parameter, but connect to a moderately old server (supports refresh tokens, but doesn't use them by default), then refresh tokens will not be used even though both the server and client support it
- Use `/_matrix/client/v4/login` for logins with refresh tokens.  Clients will know whether to use `r0/login` or `v4/login` by checking `/versions` and seeing if the server supports the spec version that introduces refresh tokens.  At some point, `r0/login` could be removed, which will break old clients, but the breakage will be noticeable rather than silent.
  - downside: clients need to check `/versions` to see which endpoint to use
- same as above, but rather than using `/versions` to signal whether to use `r0/login` or `v4/login`, add a flag to `GET /login` that indicates whether refresh tokens are supported
  - downside: clients need to check `GET /login` to see which endpoint to use, but many do that already to see whether to log in using password or SSO
- replace current login system with [UIAuth](https://github.com/matrix-org/matrix-doc/pull/2835) or [OAuth](https://github.com/matrix-org/matrix-doc/issues/2615).  Logins using the new system use refresh tokens, and logins using the old system don't.  At some point, the old system can be turned off.
  - downside: replacing the login system may take a while, and would block this from being used

I suspect that either option 2 or 3 would be the best.

Reviewer 3: Thanks for the summary, @uhoreg, that's very helpful and sounds about right to me. I agree that we need to get alignment on this point.

To be clear, I think option 1 is the current proposal, and option 4 is the long-term plan, but the idea was to get this change out as a way of breaking down the problem rather than having to deal with everything at once.

> * clients that haven't been updated will break silently when servers default to using refresh tokens.
> * [with `/v4/login`] ... the breakage will be noticeable rather than silent. 

When you say "clients will break silently": my understanding/expectation is that said clients will work fine for 5 minutes, then suddenly log you out. (Maybe we could even use the existing `soft_fail` mechanism to get clients to prompt for a reauthentication rather than a complete logout.)  

That's certainly pretty sucky UX, but it's not complete silent breakage, and from a user's point of view, it's not obviously worse than an obscure error from a client trying to use `/r0/login` - we could *try* returning an error message along the lines of "your client is too old", but I suspect most won't show it.

[maybe we should invent a time-machine so that we can go and add a `M_CLIENT_TOO_OLD` error to `/r0/login`].

Author: After discussing with @richvdh, I opted to have a `refresh_token` field in the login/registration APIs that indicate client support. Please check out the up to date "Proposal" section of the MSC

Reviewer 3: I'm going to resolve this thread, as we have a plan for compatibility now.

> # MSC2918: Refresh tokens

Reviewer 4: Overall feeling on this MSC is that it wants me to read the OAuth spec, but it hasn't given me links to that for areas of relevance. The MSC also doesn't really explain why this is needed outside of "OAuth doesn't do this" - we kinda need reference within the MSC as to why this change is important, and why we should go with OAuth practices over our own invented security principles. 

For both concerns, a list containing links to reading material are fine. For the "why oauth" bit, links to existing issues on matrix-doc about it placed in the MSC are also fine, if they help justify the MSC on their own.

Author: Makes sense, I'll clarify why we're doing this

> Requests are currently authenticated using non-expiring, revocable access tokens.

Reviewer 3: ```suggestion
In Matrix, requests to the Client-Server API are currently authenticated using non-expiring, revocable access tokens.
```

This goes against security best practices known in the OAuth 2.0 world.

> This MSC make the access tokens expiring and introduces refresh tokens to renew them to fight against token replay attacks.

Reviewer 3: ```suggestion
This MSC adds support for expiring access tokens and introduces refresh tokens to renew them.
```

## Proposal

The access token returned by the login endpoint expires after a short amount of time, forcing the client to renew it with a refresh token.
A refresh token is issued on login and rotates on each usage.

> Homeservers can choose to make the access tokens signed and non-revocable for performance reasons if the expiration is short enough (less than 5 minutes).

Reviewer 4: Signed how? What are the benefits or drawbacks? This statement feels a bit light

Reviewer 4: ah, alright. Can we get that mentioned in the sentence somewhere? Something like:

```suggestion
Homeservers can choose to make the access tokens signed and non-revocable (like a JWT, Macaroon, etc) if the expiration is short enough (less than 5 minutes). This can help with performance by avoiding DB/auth server hits.
```

### Login API changes

The login API returns two additional fields:

> - `expires_in_ms`: The lifetime in milliseconds of the access token.

Reviewer 4: we can drop this to just `expires_in` given timestamps are already in milliseconds in Matrix.

Reviewer 4: ah, fair enough. Let's keep it with `_ms` then.

> - `refresh_token`: The refresh token, which can be used to obtain new access tokens.

Reviewer 3: we should make it clear that these are *optional* fields. If they are not present, a client should assume that its `access_token` will not expire.

Author: Done in https://github.com/matrix-org/matrix-doc/pull/2918/commits/a050dc3a17cb50554ff11423d09ccdc75ae6975e

This also applies to logins done by application services.

### Account registration API changes

Unless `inhibit_login` is `true`, the account registration API returns two additional fields:

- `expires_in_ms`: The lifetime in milliseconds of the access token.

> - `refresh_token`: The refresh token, which can be used to obtain new access tokens.

Reviewer 4: Does the refresh token expire? How does one manually expire it?

Reviewer 4: also would be interested to see how this plays with soft logout: if the access token expires, but the refresh token is still live, should the server be using `soft_logout: true` in expiration responses? If so, how should the server clean up the device once the refresh token also expires?

Reviewer 4: > Refresh token don't expire, but they get invalidated on use.

This particular piece is a bit concerning, as it means that refresh tokens are hanging around waiting to give access back to the account. On the other hand, this somewhat fixes the scripts usecase as it can then store the refresh token and use that on the next run.

Author: Also clarified about `soft_logout` in the refresh token API in https://github.com/matrix-org/matrix-doc/pull/2918/commits/2c11e6f8306d45d66b28c336339172a37fb6a97a

Reviewer 4: ah ha, so the server would revoke the access token when a refresh token is used twice?

Reviewer 3: @turt2live is this clear enough now?

> This also applies to registrations done by application services.

Reviewer 4: To be clear, does this affect the masquerading of users with the `as_token` too, or just the returned token?

Reviewer 3: @sandhose: @turt2live is talking about a different mechanism of authentication for application services, which is that ASes can present a pre-shared secret known as an `as_token` in place of an `authentication_token` (see https://matrix.org/docs/spec/application_service/r0.1.2#authorization).

AFAIK we're not proposing any changes there, so the answer to travis's question is "just the returned token".

### Token refresh API

This API lets the client refresh the access token.
A new refresh token is also issued, and the existing one is revoked.

> Since this request can get lost in flight, the server should delay this revocation to when the client uses the new access token or the new refresh token for the first time.

Reviewer 3: please wrap all your lines at about 80 characters.

> A new refresh token is also issued, and the existing one is revoked.
> Since this request can get lost in flight, the server should delay this revocation to when the client uses the new access token or the new refresh token for the first time.

Reviewer 3: ```suggestion
A new refresh token is also issued. The existing refresh token remains valid until the new access token (or refresh token) is used, at which point it is revoked. This allows for the request getting lost in flight.
```

> Since this request can get lost in flight, the server should delay this revocation to when the client uses the new access token or the new refresh token for the first time.

Reviewer 4: revocation is still covered by the token expiring, correct?

Reviewer 4: Ah, alright. Reading it with those definitions makes more sense. I was thinking in the wrong headspace when reading: in code, the token would be "revoked" (deleted) when it expires, or when the user requests, which is where this caught me off guard. 

Probably more of a me problem than an MSC issue.

> The Matrix server can but does not have to make the old access token invalid, since its lifetime is short enough.

Reviewer 4: this is a bit confusing: surely the old token should be invalidated at some point regardless of methodology?

Reviewer 4: Yup, that's a lot clearer - thank you!

> `POST /_matrix/client/r0/refresh`

Reviewer 4: For appservices, I assume they want to use this without masquerading given the downstream usages of the refresh token will be unaware of their access token. Should we just go a step further and say masquerading is ignored on this endpoint given it's not super clear what that would look like?

Reviewer 4: This is based off the same [authentication clarification](https://github.com/matrix-org/matrix-doc/pull/2918#discussion_r662248851] richvdh supplied on the other thread: if an appservice has two registered users, Alice and Bob, with refresh tokens `aaa` and `bbb` then the appservice could potentially hit this API as `/refresh?user_id=@bob:localhost` with `{refresh_token: "aaa"}` (a mismatched user and token). Instead of having the homeserver try and figure out this leap of logic, we could just spec that masquerading (`?user_id=`) has no effect on this endpoint.

Banning masquerading on the endpoint should be fine because realistically no appservice should be insane enough to reauthenticate this way as they'll only have gotten the refresh token from registration or logging in, which means they don't need the additional context that masquerading would provide. If an appservice forgot the user ID a refresh token was for, it could renew it for an access token and hit `/account/whoami` as well.

> `POST /_matrix/client/r0/refresh`

Reviewer 4: Does this endpoint require authentication, and can it be rate limited?

Reviewer 3: well, it requires authentication in that you have to supply a refresh token. It's not the same sort of authentication as we use elsewhere in the API.

Reviewer 4: ah, yes, sorry: I meant authentication in the sense that we need an `Authorization` header, which does not sound like the case here.

A mention of rate limiting/authentication is largely pedantics for the spec process to make it unambiguously clear what the MSC wanted for these properties.

```json
{
  "refresh_token": "aaaabbbbccccdddd"
}
```

response:

```json
{
  "access_token": "xxxxyyyyzzz",
  "expires_in_ms": 60000,
  "refresh_token": "eeeeffffgggghhhh"
}
```

The `refresh_token` parameter can be invalid for two reasons:

 - if it does not exist
 - if it was already used once

> In both cases, the server must reply with a `401` HTTP status code and an `M_UNKNOWN_TOKEN` error code.

Author: It could also be split here, maybe having rather a `403` / `M_FORBIDDEN` (or another code?) if the token was already used

Reviewer 4: I'd recommend sticking with the 401 error to allow servers to clear out records from their database, and for mild security against attackers trying to validate if a token is/was valid at one point (proving that their data might be reliable).

This new use case of the `M_UNKNOWN_TOKEN` error code must be reflected in the spec.

### Device handling

The current spec states that "Matrix servers should record which device each access token is assigned to".
This must be updated to reflect that devices are bound to a session, which are created during login and stays the same after refreshing the token.

## Potential issues

The refresh token being rotated on each refresh is strongly recommended in the OAuth 2.0 world for unauthenticated clients to avoid token replay attacks.
This can however make the deployment of CLI tools for Matrix a bit harder, since the credentials can't be statically defined anymore.
This is not an issue in OAuth 2.0 because usually CLI tools use the client credentials flow, also known as service accounts.
An alternative would be to make the refresh token non-rotating for now but recommend clients to support rotation of refresh tokens and enforce it later on.

## Alternatives

> This MSC defines a new endpoint for token refresh, but it could also be integrated as a new authentication mechanism.

Reviewer 4: to expand on this, and the "potential issues" section above, what are the concerns with introducing it as some form of opt-in (or opt-out) mechanism for things like long-lived bots or scripts which do not easily have a refresh opportunity? For example, a nightly batch job to prune rooms/events/etc could use a static access token instead of having to login, do the work, then log out again, which would put the password near the script rather than a single revocable token.

Author: I think for both use cases (bots and scripts) I'd rather make use of the `org.matrix.login.jwt` login type with some KMS signing it for the initial login and still have the token refresh. Storing long-lived access tokens without proper secret handling is at least as bad as storing the login/pass of the bot IMO, especially if the user has admin access.
If we still need some kind of static access token, I'd rather have that in Synapse (in the config or something) than in the spec

Reviewer 4: Fair, that sounds reasonable. Just wanted to expand on the potential usecase, but agreed that scripts can find other ways to authenticate (or better yet: be replaced by features within the protocol/homeserver implementation)

Reviewer 2: I think an authentication option for scripts needs to be in the spec. I have a lot of scripts that push notifications or upload files from CI jobs for example. Those use access tokens, because CI jobs do sometimes get compromised (happened once because of codecov) and that way the access token can be easily rotated without being a homeserver admin. If the script used username and password instead, an attacker would have been able to get past UIA and change the password and just in general do much more nasty stuff than with an access token. The jobs also can't refresh the access token, since they may be running concurrently and can't change CI variables.

What would be my alternative for that use case, that works independent of the specific homeserver implementation?

Author: There is an ongoing effort to rework the whole authentication process, with use cases like scripts running in CI in mind. This MSC is also done to prepare clients for the eventual migration to this new authentication stack without having them to logout all their existing sessions.

The login API with non-expiring token will hopefully stay until this new auth stack is ready, so when you would need to migrate you will have a proper alternative.

In the meantime, if you want to still adopt refresh tokens and you are admin of your homeserver, I suggest you look into the [`org.matrix.login.jwt` login type in Synapse](https://anoadragon453.github.io/synapse/latest/jwt.html). Even though it is not standard, it will let you login using a JWT signed by some party.
It might still need some changes in Synapse to allow restricting the tokens (like not being able to use them for UIA to avoid letting the account to be nuked), but I prefer to go that route rather than adding this kind of special cases in this MSC if it will be superseded by something else soon-ish.

## Proposal

Homeservers can choose to have access tokens expire after a short amount of time, forcing the client to renew them with a refresh token.
A refresh token is issued on login and rotates on each usage.

It allows homeservers to opt for signed and non-revocable access tokens (JWTs, Macaroon, etc.) for performance reasons if their expiration is short enough (less than 5 minutes).

> Whether the access token expire and what lifetime they have is up to the homeserver, but client have to support refreshing tokens.

Reviewer 3: I think we can just get rid of this sentence.

```suggestion
```

It is heavily recommended for clients to support refreshing tokens for additional security.
They can advertise their support by adding a `"refresh_token": true` field in the request body on the `/login` and `/register` APIs.

Handling of clients that do *not* support refreshing access tokens is up to individual homeserver deployments.
For example, server administrators may choose to support such clients for backwards-compatibility, or to expire access tokens anyway for improved security at the cost of inferior user experience in legacy clients.

> If a client uses an access token that has expired, the server will respond with an `M_UNKNOWN_TOKEN` error.

Author: done here: https://github.com/matrix-org/matrix-doc/pull/2918/commits/70b2dfcde2f087c4c2e9eea33792e0b13a64e0e9

Thus, if a client receives an `M_UNKNOWN_TOKEN` error, and it has a refresh token available, it should no longer assume that it has been logged out, and instead attempt to refresh the token.

> If the client was in fact logged out, then the server will respond with an `M_UNKNOWN_TOKEN` error, possibly with the `soft_logout` parameter set.

Reviewer 3: this is confusing. It says:
 * If the client was not logged out, but the access token expires, the server responds with `M_UNKNOWN_TOKEN` error.
 * If the client *was* logged out, the server responds with `M_UNKNOWN_TOKEN` error.

so the same thing in each case? It seems to be a complicated way of saying that.

Author: The last sentence talks about the token refresh request, so it's more a "if you get a `M_UNKNOWN_TOKEN` to a random request, you should try refreshing, and if you get a `M_UNKNOWN_TOKEN` when refreshing, you're really logged out"

Reviewer 3: ah I see, that wasn't clear to me.
